torchtune/modules/layer_dropout.py

```diff
@@ -0,0 +1,302 @@
+        self.module = module
+        self.dropout = dropout
+
+    def forward(self, input: torch.Tensor, *args, **kwargs):
+        return self.dropout(self.module, input, *args, **kwargs)
+
+    def __getattr__(self, name: str) -> Any:
+        """Forward missing attributes to wrapped module."""
+        try:
+            return super().__getattr__(name)  # defer to nn.Module's logic
+        except AttributeError:
+            return getattr(self.module, name)  # fallback to wrapped module
+
+    def __setattr__(self, name: str, value: Any) -> Any:
+        """Forward missing attributes to wrapped module."""
+        try:
+            return super().__setattr__(name, value)  # defer to nn.Module's logic
+        except AttributeError:
+            return setattr(self.module, name, value)  # fallback to wrapped module
+
+    def __getitem__(self, key: int) -> Any:
+        """Forward indexing calls in case the module is a nn.Sequential."""
+        return self.module.__getitem__(key)
+
+    def state_dict(self, *args, **kwargs):
+        """Return the state dictionary of the wrapped module."""
+        return self.module.state_dict(*args, **kwargs)
+
+    def load_state_dict(self, state_dict, *args, **kwargs):
+        """Load the state dictionary into the wrapped module."""
+        self.module.load_state_dict(state_dict, *args, **kwargs)
+        return
+
+
+class ScaleType(str, Enum):
+    UNIFORM = "uniform"
+    EXP = "exp"
+    LINEAR = "linear"
+    LOG = "log"
+    SIN = "sin"
+    SIGMOID = "sigmoid"
+    STEP = "step"
+
+
+def get_scale(
+    scale_type: ScaleType,
+    scale_period: int,
+    val: int,
+) -> float:
+    """
+    Compute a scaling factor based on the provided scale type, period, and value.
+    The scaling factor is designed to be 0 when the value is 0 and 1 when the value
+    reaches or is larger than the scale period.
+    Args:
+        scale_type (ScaleType): The type of scaling to use.
+        scale_period (int): The period over which the scaling factor increases from 0 to 1.
+        val (int): The current value used to compute the scaling factor.
+    Returns:
+        float: The computed scaling factor.
+    Examples:
+        >>> get_scale(ScaleType.LINEAR, 10, 5)
+        0.5
+        >>> get_scale(ScaleType.LINEAR, 10, 0)
+        0.0
+        >>> get_scale(ScaleType.LOG, 10, 10)
+        1.0
+    """
+    if scale_period == 0:
+        return 1.0
+
+    # all the equations below aim to make scale = 0 when val=0, and scale = 1 when val=scale_period
+    scale = {
+        ScaleType.UNIFORM: 1.0,
+        ScaleType.EXP: math.pow(2, val / scale_period) - 1,
+        ScaleType.LINEAR: val / scale_period,
+        ScaleType.LOG: math.log(val + 1, scale_period + 1),
+        ScaleType.SIN: math.sin(0.5 * math.pi * val / scale_period),
+        ScaleType.SIGMOID: 1 / (1 + math.exp(-10 * (val / scale_period - 0.5))),
+    }[scale_type]
+
+    # after scale_period, scale should be 1
+    return min(scale, 1.0)
```

Comment:
At least for sin you may have to also include a `max(*, 0.0)` (though in that case for val > scale_period I guess you will start getting weird behavior anyways)